pkg/agent/agentclient/stream.go

```diff
@@ -0,0 +1,213 @@
+package agentclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"k8s.io/klog"
+	"sigs.k8s.io/apiserver-network-proxy/proto/agent"
+	// "k8s.io/client-go/util/retry"
+)
+
+const (
+	defaultRetry    = 20
+	defaultInterval = 5 * time.Second
+)
+
+type ReconnectError struct {
+	internalErr error
+	errChan     <-chan error
+}
+
+func (e *ReconnectError) Error() string {
+	return "transient error: " + e.internalErr.Error()
+}
+
+func (e *ReconnectError) Wait() error {
+	return <-e.errChan
```

Comment:
So I see code reading from the error channel but nothing writing to it?

Reply:
it's queued in `reconnWaiters`, and `doneReconnect()` writes to all the waiters channels.

Reply:
Sorry missed that. That leaves me with the question where is it closed?

Reply:
it's not closed. if it is not used, it will be gc'ed by system.